[dp-thread-pool] Add port for DeveloperPaul123/thread-pool

ptsouchlos opened the pull request:
If this PR adds a new port, please uncomment and fill out this checklist:

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

ptsouchlos committed:
Add port for DeveloperPaul123/thread-pool

DeveloperPaul123 commented:
@microsoft-github-policy-service agree

ptsouchlos committed:
Format vcpkg.json file

ptsouchlos committed:
Update baseline versions

DeveloperPaul123 commented:
I could use some help with the failing CI checks. I tried `x64-windows` and `x86-windows` locally and they both work:

```
PS D:\Tools\vcpkg> vcpkg install devpaul123-thread-pool:x86-windows
Computing installation plan...
The following packages will be built and installed:
    devpaul123-thread-pool:x86-windows -> 0.6.0
Detecting compiler hash for triplet x86-windows...
Restored 0 package(s) from C:\Users\Paul\AppData\Local\vcpkg\archives in 391 us. Use --debug to see more details.
Installing 1/1 devpaul123-thread-pool:x86-windows...
Building devpaul123-thread-pool:x86-windows...
-- Using cached DeveloperPaul123-thread-pool-0.6.0.tar.gz.
-- Cleaning sources at D:/Tools/vcpkg/buildtrees/devpaul123-thread-pool/src/0.6.0-d6b0341938.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source D:/Tools/vcpkg/downloads/DeveloperPaul123-thread-pool-0.6.0.tar.gz
-- Using source at D:/Tools/vcpkg/buildtrees/devpaul123-thread-pool/src/0.6.0-d6b0341938.clean
-- Found external ninja('1.11.0').
-- Configuring x86-windows
-- Building x86-windows-dbg
-- Building x86-windows-rel
-- Installing: D:/Tools/vcpkg/packages/devpaul123-thread-pool_x86-windows/share/devpaul123-thread-pool/copyright
-- Performing post-build validation
Stored binaries in 1 destinations in 28.9 ms.
Elapsed time to handle devpaul123-thread-pool:x86-windows: 4.4 s
Total install time: 4.4 s
devpaul123-thread-pool provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(ThreadPool CONFIG REQUIRED)
    target_link_libraries(main PRIVATE ThreadPool::ThreadPool)
```

```
PS D:\Tools\vcpkg> vcpkg install devpaul123-thread-pool
Computing installation plan...
The following packages will be built and installed:
    devpaul123-thread-pool:x64-windows -> 0.6.0
Detecting compiler hash for triplet x64-windows...
Restored 0 package(s) from C:\Users\Paul\AppData\Local\vcpkg\archives in 259 us. Use --debug to see more details.
Installing 1/1 devpaul123-thread-pool:x64-windows...
Building devpaul123-thread-pool:x64-windows...
-- Using cached DeveloperPaul123-thread-pool-0.6.0.tar.gz.
-- Cleaning sources at D:/Tools/vcpkg/buildtrees/devpaul123-thread-pool/src/0.6.0-d6b0341938.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source D:/Tools/vcpkg/downloads/DeveloperPaul123-thread-pool-0.6.0.tar.gz
-- Using source at D:/Tools/vcpkg/buildtrees/devpaul123-thread-pool/src/0.6.0-d6b0341938.clean
-- Found external ninja('1.11.0').
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Installing: D:/Tools/vcpkg/packages/devpaul123-thread-pool_x64-windows/share/devpaul123-thread-pool/copyright
-- Performing post-build validation
Stored binaries in 1 destinations in 26.8 ms.
Elapsed time to handle devpaul123-thread-pool:x64-windows: 3.8 s
Total install time: 3.8 s
devpaul123-thread-pool provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(ThreadPool CONFIG REQUIRED)
    target_link_libraries(main PRIVATE ThreadPool::ThreadPool)
```

Adela0814 commented on ports/devpaul123-thread-pool/portfile.cmake:
```diff
@@ -0,0 +1,27 @@
+)
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
+    GENERATOR "Ninja"
```
```suggestion
    SOURCE_PATH "${SOURCE_PATH}"
```

Adela0814 commented:
Please fix the CI error:
```
CMake Error at D:/downloads/tools/cmake-3.25.1-windows/cmake-3.25.1-windows-i386/share/cmake-3.25/Modules/ExternalProject.cmake:2790 (message):
  error: could not find git for clone of packageproject.cmake-populate
Call Stack (most recent call first):
  D:/downloads/tools/cmake-3.25.1-windows/cmake-3.25.1-windows-i386/share/cmake-3.25/Modules/ExternalProject.cmake:4185 (_ep_add_download_command)
  CMakeLists.txt:23 (ExternalProject_Add)
```
Please see details from: https://dev.azure.com/vcpkg/public/_build/results?buildId=91760&view=artifacts&pathAsName=false&type=publishedArtifacts

Adela0814 commented:
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

Commit:
Update ports/devpaul123-thread-pool/portfile.cmake

Co-authored-by: Mengna Li <95600143+Adela0814@users.noreply.github.com>

dg0yt commented:
> Does the container that the tests run on not have `git` installed? 

There is a way to make `git` available during the port build ...

My project auto downloads it's dependencies

... but this doesn't fit for vcpkg. You can use vcpkg ports, or you can do additional downloads from the portfile via maintainer functions.

dg0yt commented on ports/devpaul123-thread-pool/portfile.cmake:
```diff
@@ -0,0 +1,26 @@
+        -DTP_BUILD_TESTS=OFF
+        -DTP_BUILD_EXAMPLES=OFF
+        -DTP_BUILD_BENCHMARKS=OFF
+        -DFETCHCONTENT_FULLY_DISCONNECTED=OFF
```
No, thanks. Move thee downloads to the portfile at least. Let them go through vcpkg's asset caching.

dg0yt commented on ports/devpaul123-thread-pool/portfile.cmake:
```diff
@@ -0,0 +1,26 @@
+    CONFIG_PATH lib/cmake/ThreadPool-${VERSION}
+)
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/" "${CURRENT_PACKAGES_DIR}/lib")
```
Header-only ports can simply start with `set(VCPKG_BUILD_TYPE release) # header-only`, then this is reduced to
```suggestion
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib")

```

dg0yt commented on ports/devpaul123-thread-pool/portfile.cmake:
```diff
@@ -0,0 +1,26 @@
+
+vcpkg_cmake_install()
+vcpkg_cmake_config_fixup(
+    PACKAGE_NAME ThreadPool
```
You can use lower-case `threadpool` here. It is just the directory name. On case-sensitive file system, it will exist anyways.

Adela0814 commented:
> @Adela0814 Thanks for the info.
> 
> Does the container that the tests run on not have `git` installed? My project auto downloads it's dependencies and it relies on `git` being available. The error seems to be caused by git not being found:
> 
> ```
> CMake Error at D:/downloads/tools/cmake-3.25.1-windows/cmake-3.25.1-windows-i386/share/cmake-3.25/Modules/ExternalProject.cmake:2790 (message):
>   error: could not find git for clone of packageproject.cmake-populate
> Call Stack (most recent call first):
>   D:/downloads/tools/cmake-3.25.1-windows/cmake-3.25.1-windows-i386/share/cmake-3.25/Modules/ExternalProject.cmake:4185 (_ep_add_download_command)
>   CMakeLists.txt:23 (ExternalProject_Add)
> ```

`git` is installed on the test machine.

DeveloperPaul123 commented:
@dg0yt

> There is a way to make git available during the port build ...

Why specifically during the port build? Plus @Adela0814 mentioned that git is installed. Maybe there's something I'm missing....

> ... but this doesn't fit for vcpkg. You can use vcpkg ports, or you can do additional downloads from the portfile via maintainer functions.

The project is just downloading a cmake script to make packaging the project easier. Thus this is not available via `vcpkg` ports. I guess I will have to try using the vcpkg maintainer functions and move this to the port file.

ptsouchlos committed:
Rename port to dp-thread-pool

Updated commit hash (upstream project fixed some cmake issues for vcpkg support)

ptsouchlos committed:
vcpkg x-add-version update

ptsouchlos committed:
Fix packagename and config path

ptsouchlos committed:
Use tagged version of library

dg0yt commented on ports/dp-thread-pool/portfile.cmake:
```diff
@@ -0,0 +1,30 @@
+    HEAD_REF master
+)
+
+set(PACKAGE_PROJECT_CMAKE_LOCATION "${SOURCE_PATH}/3rdparty/PackageProject.cmake")
```
Where is this variable used?

DeveloperPaul123 replied:
It's not, I'll remove it

dg0yt reviewed (commented):
IIUC there is a hidden unconditional download of CPM, https://github.com/DeveloperPaul123/thread-pool/blob/d6a49b6cb5c24fcac83f51d372f1396ae0faa89a/cmake/CPM.cmake#L13C18-L19.
Is it needed/used?

ptsouchlos committed:
Remove unused variable

DeveloperPaul123 commented:
> IIUC there is a hidden unconditional download of CPM, https://github.com/DeveloperPaul123/thread-pool/blob/d6a49b6cb5c24fcac83f51d372f1396ae0faa89a/cmake/CPM.cmake#L13C18-L19. Is it needed/used?

You are correct, I will fix this upstream.

ptsouchlos committed:
Update to commit that doesn't blindly download CPM.cmake

DeveloperPaul123 commented:
I addressed the unconditional CPM.cmake download

Adela0814 commented on ports/dp-thread-pool/portfile.cmake:
```diff
@@ -0,0 +1,28 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO DeveloperPaul123/thread-pool
+    REF 517e39f02207c4a0d227c77f0bc1cfb0518044be
```
```suggestion
    REF ${VERSION}
```

DeveloperPaul123 replied:
The commit doesn't exactly match the tagged release since I had to patch the CPM download issue. Do I need to tag a new release upstream?

Adela0814 commented:
Usage tested pass on `x64-windows`.

ptsouchlos committed:
Update portfile and version

ptsouchlos committed:
Update versions
